doc/designs/quic-design/error-handling.md

```diff
@@ -0,0 +1,101 @@
+APIs with the same SSL object from different threads, but even if it happens,
+it is not a problem as applications are supposed to check for errors
+immediately after the API call on the same thread. There is no such thing as
+Thread-assisted mode of operation.
+
+Constraints
+-----------
+
+We need to keep using the existing ERR API as doing otherwise would
+complicate the existing applications and break our API compatibility promise.
+Even the ERR_STATE structure is public, although deprecated, and thus its
+structure and semantics cannot be changed.
+
+The error stack access is not under a lock (because it is thread-local).
+This complicates _moving errors between threads_.
+
+Error stack entries contain allocated data, copying entries between threads
+would mean duplicating it or losing it.
+
+Assumptions
+-----------
+
+This document assumes the error state of the QUIC connection (or stream for
+stream level errors) is handled separately.
+
+We can assume the internal assistance thread is well-behaving in regards
+to the error stack.
+
+We assume there are two types of errors that can be raised in the QUIC
+library calls and in the subordinate libcrypto (and provider) calls. First
+type is an intermittent error that does not really affect the state of the
+QUIC connection - for example EAGAIN returned on a syscall, or unavailability
+of some algorithm where there are other algorithms to try. Second type
+is a permanent error that affects the error state of the QUIC connection.
+Operations on QUIC streams (SSL_write(), SSL_read()) can also trigger errors,
+depending on their effect they are either permanent if they cause the
+QUIC connection to enter an error state, or if they just affect the stream
+they are left on the error stack of the thread that called SSL_write()
+or SSL_read() on the stream.
+
+Design
+------
+
+Return value of SSL_get_error() on QUIC connections or streams does not
+depend on the error stack contents.
+
+Intermittent errors are handled within the library and cleared from the
+error stack before returning to the user.
+
+Permanent errors happenning within the assist thread need to be transferred
+to the regular user thread. To simplify the implementation they are required
+to appear on the error stack reachable by the application ERR_ calls only
+when the SSL_get_error() is called from the application.
+
+Implementation
+--------------
+
+There is an error stack in QUIC_CHANNEL which serves as temporary storage
+for errors happening in the internal assistance thread. When a permanent error
+is detected in the assistance thread the error stack entries are copied
+to this error stack in QUIC_CHANNEL.
+
+When SSL_get_error() is called and there is a permanent error condition
+detected the errors from the QUIC_CHANNEL error stack are moved to the
+error stack of the thread that is calling SSL_get_error() on the SSL
+object representing the QUIC_CONNECTION.
```

Comment:
> I agree with storing an error stack inside the `QUIC_CHANNEL` which captures the failure reason. But I'm not sure `SSL_get_error` makes sense as an API to query it? SSL_get_error hasn't been an API to _mutate_ the error stack before, it's been an API to query various normal and non-normal error conditions, some of which depend on the current state of the error stack.

Hmm, that's a valid concern. I wanted to avoid doing the err stack copying at various random places, but you're right that doing that in SSL_get_error() could subtly break some applications.

> We could have a new API to explicitly replace the current thread-local ERR state with the stored ERR which caused connection failure. However, if the concern is for compatibility with existing code, I think it would be reasonable for attempts to call API functions which fail due to the connection being terminated also raise these errors.
> 
> In other words, a typical function like `SSL_do_handshake` or `SSL_write` would go something like:
> 
> ```c
> {
>   if (connection terminated) {
>     copy stashed error stack to local thread, return 0
>      // NOTE: currently we have a RAISE_NON_NORMAL_ERROR macro here which signals failure due to
>      // connection termination. For simplicity, we can have an invariant that if the connection is in the
>      // terminated state, a non-empty error stack in the channel is guaranteed to exist.
>      // So all current if (terminated) return NON_NORMAL_ERROR() code can be replaced with some new
>      // macro, maybe RAISE_TERMINATED_ERROR() or similar. Could optionally add one extra ERR entry
>      // to the copied stack logging the thread-local post-termination call site.
>   }
>   ...
> }
> ```
> 
> But I am wary of making SSL_get_error actually generate ERRs.

Hmm... right. But how should we handle clearing the QUIC_CHANNEL error stack? Would the copying be done just once? Or, if not, there are some other concerns. We cannot know for sure whether the application cleared the previous copy before calling some other API call again. We can only approximately infer it if we look at the last reason code raised and compare it with what we have in the QUIC_CHANNEL stack.

Reply:
Ah, I see the problem. My thinking was once we entered the terminated state, the stashed channel error becomes immutable and is never changed. I think what you're raising here is a desire to avoid copying the error (or referenced strings) for performance reasons. So there is a desire for 'move' semantics rather than 'copy' here.

I guess one option is to have `SSL_write` option fail with a 'generic' error and provide some error like `SSL_get_clear_terminate_error` to 'move' the error to the thread-local error stack. Or we could just copy and accept the performance cost. Best of both worlds might be a hybrid where SSL_write raises a 'generic' error and have a `SSL_get_terminate_error` which copies &mdash; idea being that the application only calls this if it cares about the details (which are primarily diagnostic in nature and such failures should maybe not be a performance hotspot &mdash; if they are for a given application, they can just avoid calling it in production). This feels a bit more intuitive than actually forcing the error to be cleared to query it / not allowing the application to query the error more than once.

But I'm just throwing out ideas and open to other ideas.